File: packages/xandr_android/android/src/main/kotlin/de/thekorn/xandr/models/MultiAdRequestRegistry.kt
package de.thekorn.xandr.models

import com.appnexus.opensdk.ANMultiAdRequest
import com.appnexus.opensdk.Ad

val charPool: List<Char> = ('a'..'z') + ('A'..'Z') + ('0'..'9')

object MultiAdRequestRegistry {
    private val multiAdRequests = mutableMapOf<String, ANMultiAdRequest>()
    private fun generateRandomStringId() = List(16) { charPool.random() }.joinToString("")

    fun initNewRequest(mar: ANMultiAdRequest): String {
        val id = generateRandomStringId()
        multiAdRequests[id] = mar
        return id
    }

    fun removeRequestWithId(requestId: String): Boolean {
        val mar = multiAdRequests.remove(requestId)
        mar?.stop()
        return mar != null
    }

    fun load(requestId: String): Boolean {
        val mar = multiAdRequests[requestId] ?: return false
        return mar.load()
    }

    fun addAdUnit(requestId: String, ad: Ad): Boolean {
        val mar = multiAdRequests[requestId] ?: return false
        return mar.addAdUnit(ad)
    }
}

File: packages/xandr_ios/ios/Classes/models/MultiAdRequestRegistry.swift
import AppNexusSDK
import Foundation

public class MultiAdRequestRegistry {
  private var multiAdRequests: [String: ANMultiAdRequest] = [:]

  private func generateRandomStringId() -> String {
    let length = 16
    let letters = "abcdefghijklmnopqrstuvwxyzABCDEFGHIJKLMNOPQRSTUVWXYZ0123456789"
    return String((0 ..< length).map { _ in letters.randomElement()! })
  }

  public func initNewRequest(_ mar: ANMultiAdRequest) -> String {
    let id = generateRandomStringId()
    multiAdRequests[id] = mar
    return id
  }

  public func removeRequestWithId(_ requestId: String) -> Bool {
    let mar = multiAdRequests[requestId]
    mar?.stop()
    multiAdRequests[requestId] = nil
    return mar != nil
  }

  public func load(_ requestId: String) -> Bool {
    let mar = multiAdRequests[requestId]
    return mar?.load() ?? false
  }

//  public func addAdUnit(requestId: String, ad: AdUnits) -> Bool {
//    let mar = multiAdRequests[requestId]
//    return mar?.addAdUnit(ad) ?? false
//  }
}
